rotkehlchen/chain/ethereum/modules/xdai_bridge/decoder.py

```diff
@@ -0,0 +1,108 @@
+from typing import TYPE_CHECKING, Any
+
+from rotkehlchen.accounting.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.chain.evm.decoding.constants import XDAI_BRIDGE_CPT_DETAILS
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails, EventCategory
+from rotkehlchen.chain.evm.decoding.utils import bridge_match_transfer, bridge_prepare_data
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.chain.gnosis.constants import CPT_XDAI_BRIDGE
+from rotkehlchen.constants.assets import A_DAI
+from rotkehlchen.types import ChainID, ChecksumEvmAddress, DecoderEventMappingType
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+BRIDGE_ADDRESS = string_to_evm_address('0x4aa42145Aa6Ebf72e164C9bBC74fbD3788045016')
+BRIDGE_TOKEN = b'\x1dI\x1aB}\x1f\x8c\xc0\xd4GIo0\x0f\xac9\xf70a"H\x1d\x8ef4Q\xeb&\x82t\x14k'
+WITHDRAWAL_TOKEN = b'J\xb7\xd5\x813m\x92\xed\xbe\xa2&6\xa6\x13\xe8\xe7l\x99\xac\x7f\x91\x13|\x15#\xdb8\xdb\xfb;\xf3)'  # noqa: E501
+
+
+class XdaiBridgeDecoder(DecoderInterface):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.dai = A_DAI.resolve_to_crypto_asset()
+
+    def _decode_bridge_tokens(self, context: DecoderContext) -> DecodingOutput:
+        """Decodes a bridging event for dai. Either a deposit or a withdrawal."""
+        if context.tx_log.topics[0] not in {BRIDGE_TOKEN, WITHDRAWAL_TOKEN}:
```

Comment:
The logic seems really similar to the one in the gnosis chain all the `bridge_prepare_data` and the for loop logic is identical I would say with the exception on the direction. I believe you can abstract this class (maybe move it to the gnosis decoder side in a new file and then create subclasses for ethereum and gnosis.

Things that change are:

- the topics that will be allowed (this can be an attribute of the subclass
- you can create a method to extract the `from_address/to_address`
- the `main_chain`/`l2_chain` can be attributes of the subclass (I believe you only need to provice as argument the l2_chain since the main_chain is in the inquirer.chain_id or something like this)